File: elements/flowInvoker.js
// elements/flowInvoker.js

class FlowInvoker {
    constructor(data, usedVariables) {
        this.data = data;
        this.usedVariables = usedVariables || [];
        this.maxRetries = 30;        // Max 30 retries = 60 seconds
        this.retryDelay = 2000;      // 2 seconds between retries
    }

    async makeRequest(variables) {
        const url = this.data.url
        let body = null;

        if (this.data.body) {
            body = this._replaceVariablesInObject(this.data.body, variables);
        }

        // Make initial request
        const initialResponse = await this._makeHttpRequest(url, body);

        // Check if we need to poll for final result 
        if (typeof initialResponse === 'string' && (initialResponse.startsWith("https://"))) {
            return this._pollForResult(initialResponse);
        }
        // error case
        else{
            return initialResponse
        }

    }

    async _pollForResult(resultUrl) {
        let retries = 0;

        while (retries < this.maxRetries) {
            await new Promise(resolve => setTimeout(resolve, this.retryDelay));

            try {
                const response = await this._makeHttpRequest(resultUrl);

                // Check if we still need to wait
                if (typeof response === 'string' &&
                    response.includes('Processing...')) {
                    retries++;
                    continue;
                }

                // We have the final result
                return response;
            } catch (error) {
                console.error(`Polling attempt ${retries + 1} failed:`, error);
                retries++;
            }
        }

        throw new Error('Max polling attempts reached without final result');
    }

    async _makeHttpRequest(url, body = null) {
        // Prepare headers
        const headers = {};

        // Handle authentication
        if (this.data.authentication) {
            switch (this.data.authentication.type) {
                case 'Bearer':
                    headers['Authorization'] = `Bearer ${this._replaceVariables(
                        this.data.authentication.token
                    )}`;
                    break;
                case 'Basic':
                    const username = this._replaceVariables(
                        this.data.authentication.username
                    );
                    const password = this._replaceVariables(
                        this.data.authentication.password
                    );
                    const credentials = Buffer.from(
                        `${username}:${password}`
                    ).toString('base64');
                    headers['Authorization'] = `Basic ${credentials}`;
                    break;
            }
        }
        // Prepare request options (if body is null the the request type is GET)
        const options = {
            method: body ? 'POST' : 'GET',
            headers
        };

        // Add body if exists
        if (body) {
            options.body = JSON.stringify(body);
            headers['Content-Type'] = 'application/json';
        }

        // Make the request
        try {
            const response = await fetch(url, options);

            if (!response.ok) {
                const text = await response.text();
                throw new Error(`HTTP error! status=${response.status}, body=${text}`);
            }

            // should return the (/api/webhook/result) url
            return await response.text();
        } catch (error) {
            console.error('HTTP request failed:', error);
            throw error;
        }
    }

    _replaceVariables(template, variables) {
        if (typeof template !== 'string') return template;

        return template.replace(/\$\{(\w+)\}/g, (match, varName) => {
            return this.usedVariables.includes(varName)
                ? (variables[varName] || '')
                : match;
        });
    }

    _replaceVariablesInObject(obj, variables) {
        // Recursively replace variables in object properties
        if (typeof obj === 'object' && obj !== null) {
            const newObj = Array.isArray(obj) ? [] : {};
            for (const key in obj) {
                newObj[key] = this._replaceVariablesInObject(obj[key], variables);
            }
            return newObj;
        } else if (typeof obj === 'string') {
            return this._replaceVariables(obj, variables);
        }
        return obj;
    }
}

module.exports = FlowInvoker;

File: elements/flow_invoker.py
# flow_invoker.py
import json
import aiohttp
import asyncio

class FlowInvoker:
    def __init__(self, data, used_variables):
        self.data = data
        self.used_variables = used_variables
        self.max_retries = 30
        self.retry_delay = 2
    
    async def make_request(self, variables):
        url = self._replace_variables(self.data['url'], variables)
        body = None
        
        if 'body' in self.data:
            body = self._replace_variables_in_object(self.data['body'], variables)
        
        # Make initial request
        initial_response = await self._make_http_request(url, body)
        
        # Handle polling if needed
        if isinstance(initial_response, str) and initial_response.startswith('https://'):
            return await self._poll_for_result(initial_response)
        return initial_response
    
    async def _poll_for_result(self, result_url):
        retries = 0
        while retries < self.max_retries:
            await asyncio.sleep(self.retry_delay)
            try:
                response = await self._make_http_request(result_url)
                
                # Check if processing is complete
                if 'Processing...' not in response:
                    return response
            except Exception as e:
                print(f'Polling attempt {retries+1} failed: {e}')
            retries += 1
        raise Exception('Max polling attempts reached')
    
    async def _make_http_request(self, url, body=None):
        headers = {}
        method = 'POST' if body else 'GET'
        
        # Handle authentication
        if 'authentication' in self.data:
            auth = self.data['authentication']
            if auth['type'] == 'Bearer':
                token = self._replace_variables(auth['token'], {})
                headers['Authorization'] = f'Bearer {token}'
            elif auth['type'] == 'Basic':
                username = self._replace_variables(auth['username'], {})
                password = self._replace_variables(auth['password'], {})
                import base64
                credentials = base64.b64encode(f'{username}:{password}'.encode()).decode()
                headers['Authorization'] = f'Basic {credentials}'
        
        async with aiohttp.ClientSession() as session:
            async with session.request(
                method, 
                url, 
                json=body, 
                headers=headers
            ) as response:
                response.raise_for_status()
                return await response.text()
    
    def _replace_variables(self, template, variables):
        if not isinstance(template, str):
            return template
        
        for var_name in self.used_variables:
            placeholder = f'${{{var_name}}}'
            if placeholder in template:
                value = str(variables.get(var_name, ''))
                template = template.replace(placeholder, value)
        return template
    
    def _replace_variables_in_object(self, obj, variables):
        if isinstance(obj, dict):
            return {k: self._replace_variables_in_object(v, variables) for k, v in obj.items()}
        elif isinstance(obj, list):
            return [self._replace_variables_in_object(item, variables) for item in obj]
        elif isinstance(obj, str):
            return self._replace_variables(obj, variables)
        return obj